fix(stark-ui): add support for `onClickCallback` column properties in Table component

SuperITMan opened the pull request:
ISSUES CLOSED: #1466

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/NationalBankBelgium/stark/blob/master/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[X] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Cell click not handled.

Issue Number: #1466 


## What is the new behavior?
Cell click is now handled.

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

SuperITMan commented:
The tests have not been implemented yet.
Could you please have a look on the implementation I've done for `cellClicked` output in the StarkTableColumnComponent ? :blush:

christophercr commented:
And of course... unit tests 😄

christophercr commented:
Question: is this a feature or a bug? Cause #1466 is marked as bug but in your commit message it is defined as a feature

SuperITMan commented:
@christophercr
Indeed, I didn't see that I broke the row click 😛 Good catch!

It is solved now and I added 2 tests for this. Could you please check this ?
I also changed the commit message for `fix` instead of `feat` since it is a missing feature.

We maybe should invest later to write some tests for the StarkTableColumnComponent... What do you think ? 😊

christophercr commented on packages/stark-ui/src/modules/table/components/table.component.spec.ts:
```diff
@@ -1327,19 +1328,50 @@ describe("TableComponent", () => {
+	
+		it("should trigger 'onClickCallback' property when click on the cell and should not emit on 'rowClicked' when 'onClickCallback' is defined",() => {
+			const descriptionColumnElement = hostFixture.nativeElement.querySelector(
+				"table tbody tr td.mat-column-description"
```
Can you add the same check that we have in other tests to ensure the element was found?

```
if (!descriptionColumnElement) {
	fail("No description column element found");
	return;
}
```

well.. in fact the check can be re-written to `expect(descriptionColumn).not.toBeNull();`

christophercr commented on packages/stark-ui/src/modules/table/components/table.component.spec.ts:
```diff
@@ -1327,19 +1328,50 @@ describe("TableComponent", () => {
+			// click on the cell
+			triggerClick(descriptionColumnElement);
+			
+			// We expect "2" due to the check "columnProperties.onClickCallback instanceof Function" in table.component.ts
```
Strange that the callback is called twice... what if you try to move the calls of `someSpy.calls.reset()` from the `beforeEach` to one line before the `triggerClick()` call?

SuperITMan replied:
When `columnProperties.onClickCallback instanceof Function` is triggered, it generates a call.

christophercr commented on packages/stark-ui/src/modules/table/components/table.component.spec.ts:
```diff
@@ -1327,19 +1328,50 @@ describe("TableComponent", () => {
+
+		it("should not trigger 'onClickCallback' property when click on the cell but should emit on 'rowClicked' when 'onClickCallback' is not defined", () => {
+			const idColumnElement = hostFixture.nativeElement.querySelector(
+				"table tbody tr td.mat-column-id"
```
Same remark about adding a check to ensure that the `idColumnElement` was found

SuperITMan commented on showcase/src/assets/examples/table/regular/table.ts:
```diff
@@ -30,7 +30,13 @@ export class TableRegularComponent {
+			name: "description",
+			label: "Description",
+			onClickCallback: (value: any, row?: object, columnName?: string): void => {
+				console.log("CELL CLICKED - value:", value, ", row: ", row, ", columnName: ", columnName);
```
True... I'll use the logger.debug :blush:
If I use a string template, the object row will be displayed properly ?

christophercr replied:
Ah true, it might not be displayed correctly... so then keep it as it is now ;)

SuperITMan commented on showcase/src/app/demo-ui/components/table-regular/table-regular.component.ts:
```diff
@@ -62,7 +68,10 @@ export class TableRegularComponent {
-		columns: [{ columnName: "id", filterPosition: "below" }, { columnName: "title", filterPosition: "above" }],
+		columns: [
+			{ columnName: "id", filterPosition: "below" },
+			{ columnName: "title", filterPosition: "above" }
```
Yes, I do that

SuperITMan commented on packages/stark-ui/src/modules/table/components/table.component.ts:
```diff
@@ -1123,4 +1138,14 @@ export class StarkTableComponent extends AbstractStarkUiComponent implements OnI
+	private isColumnPropertiesWithOnClickCallback(
+		columnProperties?: StarkTableColumnProperties
+	): columnProperties is StarkTableColumnProperties & Required<Pick<StarkTableColumnProperties, "onClickCallback">> {
+		return !!columnProperties && columnProperties.onClickCallback instanceof Function;
```
For me, it's ok also :blush: 
I check this!

SuperITMan replied:
Actually it's like the issue I had before, it seems that I need to use a type guard in this case...
I get the following error when I use your suggestion:
```bash
Type 'StarkTableColumnProperties | undefined' is not assignable to type 'StarkTableColumnProperties & Required<Pick<StarkTableColumnProperties, "onClickCallback">>'.
  Type 'undefined' is not assignable to type 'StarkTableColumnProperties & Required<Pick<StarkTableColumnProperties, "onClickCallback">>'.
    Type 'undefined' is not assignable to type 'StarkTableColumnProperties'.ts(2322)
```

SuperITMan commented:
Changes applied. Could you please have a look @christophercr @nicanac ? :blush:
